[RT0118] iP

RT0118 opened the pull request:
**FAUNA**

> "Konfauna~ Your personalised chatbot kirin is here!" - Fauna ([👀](http://www.youtube.com/@niminightmare))

Fauna frees your mind of having to remember the things you need to do in life. It's, 
- text-based
- easy to learn
- _kinda(?)_ fast to use

All you need to do is, 
1. download [Fauna]("https://github.com/RT0118/ip/releases")
2. run it ~~with administrator privileges~~
3. add your tasks
4. profit

Features include:
- [x] Manage tasks, deadlines and events
- [ ] Manage deadlines (not yet)
- [ ] Reminders (not yet)

I still need a code block and `inline code`, so here they are:
```java
public static void main(String[] args) {
    System.out.println("lmao");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

RT0118 committed:
Rename main class and change logo

RT0118 committed:
Add greeting messages

RT0118 committed:
Add echo feature

RT0118 committed:
Implement add and list features

RT0118 committed:
Add mark and unmark task as done features

RT0118 committed:
Add ToDo, Deadline and Event classes

RT0118 committed:
Fix EventTask's incorrect identifier when printing

RT0118 committed:
Fix unnecessary space in printAddTaskPrompt, and automated text UI testing

RT0118 committed:
Add custom exceptions, handled exceptions from user input and array out of bounds, and abstracted user input parsing

RT0118 committed:
Add delete task feature and fix printing bugs

RT0118 committed:
Move bot commands to FaunaCommand enum, and add ui test cases

RT0118 committed:
rename mark as done commands

RT0118 committed:
Change Event to use from and to, change parser to use regex, and modify ui tests

RT0118 committed:
Add serialization methods for Task subclasses to save and restore data

RT0118 committed:
Create Storage class to handle saving and restoring tasks from file

RT0118 committed:
Add storage features to Fauna to save and restore tasks

RT0118 committed:
update Fauna to use a different save folder location

RT0118 committed:
Merge pull request #1 from RT0118/branch-Level-7

Merge branch-Level-7

RT0118 committed:
Implement parser methods to parse task's datetime from user input

RT0118 committed:
Update datetime parser in UserInputParser to handle user input errors

RT0118 committed:
Add date and time feature to Task classes

RT0118 committed:
Modify methods that create new Tasks objects to handle LocalDateTime

RT0118 committed:
Change gitignore to remove save file

RT0118 committed:
Merge pull request #2 from RT0118/branch-Level-8

Merge branch-Level-8

RT0118 committed:
Refactor all class files to divide into packages

RT0118 committed:
Modify command descriptions to cleanup refactor mistakes

RT0118 committed:
Merge branch 'branch-A-Packages'

RT0118 committed:
Create Ui class to handle user interactions

RT0118 committed:
Create TaskList class to abstract task list operations

RT0118 committed:
Change parser to not change task's index from user input

RT0118 committed:
Fix imports to remove auto-collapsed imports

RT0118 committed:
Update text-ui-test to add updated test cases

RT0118 committed:
Merge branch 'branch-A-MoreOOP'

RT0118 committed:
Merge branch 'add-gradle-support'

RT0118 committed:
Add config files for checkstyle

RT0118 committed:
Modify build.gradle to add checkstyle plugin

RT0118 committed:
Merge branch 'branch-A-Gradle'

RT0118 committed:
Add JUnit tests for TaskList and UserInputParser

RT0118 committed:
Merge branch 'A-JUnit'

RT0118 committed:
Remove unused import

RT0118 committed:
Add javadoc to exceptions package

RT0118 committed:
Add javadocs to parser package

RT0118 committed:
Add javadocs for storage package

RT0118 committed:
Add javadocs for task package

RT0118 committed:
Add javadocs for ui package

RT0118 committed:
Merge branch 'A-JavaDoc'

RT0118 committed:
Fix codestyle and imports

RT0118 committed:
Merge branch 'A-CodingStandard'

RT0118 committed:
Implement the find command to search for tasks using keywords

RT0118 committed:
Make Fauna more OOP

RT0118 committed:
Update ui to print command help with invalid command given

RT0118 committed:
Fix spacing issue when printing command description

RT0118 committed:
Merge branch 'branch-Level-9'

RT0118 committed:
Update build.gradle for release

RT0118 committed:
Merge branch 'branch-A-Jar'

zacklow28 commented on src/main/java/fauna/Fauna.java:
```diff
@@ -0,0 +1,163 @@
+        this.taskList = this.storage.restore();
+    }
+
+    private void addTodoToTaskList(String taskName) {
```
The add__ToTaskList methods can be placed in the TaskList class for better OOP 😃

RT0118 replied:
Oh, good suggestion! Thanks for the input

zacklow28 commented on src/main/java/fauna/parser/FaunaCommand.java:
```diff
@@ -0,0 +1,88 @@
+ * extracting variables from user input
+ */
+public enum FaunaCommand {
+    LIST("Lists all tasks"),
```
Like the ENUMS, I think they really help in code reuse!

RT0118 replied:
Thanks for the input!

zacklow28 commented on src/main/java/fauna/parser/FaunaCommand.java:
```diff
@@ -0,0 +1,88 @@
+     * to extract information from user input
+     * </p>
+     * @return regex pattern string
+     */
```
Interesting use of regular expressions!

zacklow28 commented on src/main/java/fauna/parser/ParsedUserInput.java:
```diff
@@ -0,0 +1,92 @@
+ */
+public class ParsedUserInput {
+    private final FaunaCommand command;
+    private final Optional<String> taskName;
```
Is the Optional class necessary here? How does it work with your code?

RT0118 replied:
My thought process was to return some kind of meaningful value (which in this case is "invalid ..."), instead of purely returning null. But I do agree, I could have put more thought into using Optionals here.

Correction: I have an OrElseThrow() when you query for something that doesnt exist, not return a string "invalid ..."

zacklow28 commented on src/main/java/fauna/ui/Ui.java:
```diff
@@ -0,0 +1,160 @@
+    private static final String CHATBOT_LOGO = """
+             _____ _   _   _ _   _    _    
+            |  ___/ \\ | | | | \\ | |  / \\   
+            | |_ / _ \\| | | |  \\| | / _ \\  
```
Wow looks good, good effort! 😮

RT0118 replied:
Thanks!

zacklow28 reviewed (commented):
LGTM. Overall demonstrated good use of quite a few Java concepts which I can learn from

Kannan171 commented on src/main/java/fauna/task/DeadlineTask.java:
```diff
@@ -0,0 +1,58 @@
+        this.by = by;
+    }
+
+    public DeadlineTask(String taskName, LocalDateTime by) {
```
Perhaps "by" can be a vague variable name and could be better represented as "deadline" or "doBy"

Kannan171 reviewed (commented):
LGTM. Well named and clear to read code

RT0118 committed:
Fix some issues identified by CheckStyleMain

RT0118 committed:
Merge branch 'branch-A-CheckStyle'

RT0118 committed:
Update build.gradle to add JavaFX dependencies

RT0118 committed:
Implement JavaFX features to enable GUI chatbot

RT0118 committed:
Modify Fauna and Ui to move system prints to Fauna and allow GUI to get responses

RT0118 committed:
Upload chatbot images

RT0118 committed:
Modify build.gradle to launch GUI chatbot

RT0118 committed:
Add javadocs for class files to conform to checkStyleMain

RT0118 committed:
Merge branch 'branch-Level-10'

RT0118 committed:
Implement assertions to document assumptions in code

There are certain places in code that need some variables to be
set to function properly.

Creating assumptions will help to ensure important assumptions
hold.

Let's use assert to create assumptions.

RT0118 committed:
Update build.gradle to change archiveVersion of fat JAR file

RT0118 committed:
Improve code structure to conform to coding standards

Some methods have multiple responsibilities, which reduces
readability and maintainability. This makes it harder to
maintain the codebase.

Let's,
* Abstract welcome message from Fauna
* Abstract string formatting from TaskList find command

This ensures that each method operates at a consistent
level of abstraction, making it easier to read and maintain

RT0118 committed:
Update TaskList's removeTask to make streams more readable

RT0118 committed:
Merge branch 'branch-A-Assertions'

RT0118 committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/fauna/ui/Fauna.java

RT0118 committed:
Merge branch 'branch-A-Streams'

RT0118 committed:
Add tag command to command list

RT0118 committed:
Update parser to parse tag commands

RT0118 committed:
Implement tagging capability in Tasks to add tags to tasks

RT0118 committed:
Update Ui and Fauna to handle tagging and printing tag events

RT0118 committed:
Merge pull request #4 from RT0118/C-Tagging

Merge branch 'branch-C-Tagging'

RT0118 committed:
Update Main and Fauna to comply fix code quality issues

RT0118 committed:
Merge pull request #5 from RT0118/branch-FixCodeQuality

Merge branch 'branch-FixCodeQuality'

RT0118 committed:
Add documentation for github pages

RT0118 committed:
Merge branch 'branch-A-UserGuide'

RT0118 committed:
Merge branch 'master' of https://github.com/RT0118/ip

RT0118 committed:
Update build.gradle to change fat jar file version
